teps/0033-tekton-feature-gates.md

```diff
@@ -145,6 +160,393 @@ alpha features. This is not a requirement for Tekton, because even in case of
+
+_We should avoid adding features like this if possible since each new feature
+complicates the combination of settings we'll need to test in order to feel
+confident in our automated testing, and because promoting these changes from
+alpha will be a backwards incompatible change._ 
+
+Pros:
+* It will be clear to users whether they have opted into this functionality or not
+
+Cons:
+* Flags added this way can only be removed according to our deprecation policy and will need to
+  be supported for as long as the corresponding version is supported (e.g. today we support v1alpha1
+  and v1beta1 so flags of this kind that we added for v1beta1 functionality will stick around until
+  we completely remove support for v1beta1)
+
+### Promotion to beta and beyond
+
+It will be at the discretion of [the pipelines OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS_ALIASES)
+to decide when a feature is ready to move from alpha to beta and finally to v1.
+
+This will likely require the feature being available for at least one release
+if not several releases (and maybe use in [dogfooding](https://github.com/tektoncd/plumbing/blob/main/docs/dogfooding.md))
+to be able to collect and handle user feedback.
+
+* When an alpha API field is ready to be declared beta:
+    * If the underlying CRD has a version of v1, the webhook admission controller
+      will be updated to require the `enable-beta-api-fields` flag instead of the
+      `enable-alpha-api-fields` flag in order to use the field.
+    * If the underlying CRD has a version of v1beta1, the webhook admission controller
+      will be updated no longer require `enable-alpha-api-fields` to use the field
+* When a beta API field is ready to be declared v1, the webhook addmission controller
+  will be updated to no longer require `enable-beta-api-fields` to use the field.
+
+#### Changes in behavior
+
+Promoting the functionality guarded by a specific feature flag to beta and beyond would
+mean enabling it by default and would be a breaking change.
+
+This will mean following [the API compatibility policy](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) for all
+impacted CRDs.
+
+For example, we currently have a feature flag `disable-working-directory-overwrite` which was added
+as part of a deprecation announcement in
+[Tekton Pipelines v0.11.0](https://github.com/tektoncd/pipeline/releases/tag/v0.11.0-rc1) which was
+Mar 4, 2020, and was our first beta release. Our
+[API compatibility policy for beta](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#alpha-beta-and-ga)
+requires that we give at least 9 months to migrate from the depcrecated functionality, so
+[as early as Dec 4, 2020](https://github.com/tektoncd/pipeline/blob/master/docs/deprecations.md)
+we have the option to either change the default value of the flag (and later remove it) or to remove
+the flag completely.
```

Comment:
@afrittoli you mentioned we should add an example so I've added the above, let me know if that does the trick

Reply:
Thank you, this works!
We should decide as a result of this TEP whether we changing the default value first is an option.

We could split up the 6 months into 3 deprecation and feature flag that defaults to the deprecated behaviour plus 3 months with feature flag that defaults to the new behaviour, and Tekton being very vocal in logs if someone flicks the switch back to enabled.

For the case of the working directory overwrite specifically for instance we might want to switch the default value asap and add verbouse warnings with a deadline for removal.

Reply:
hm hm hm it still feels like a backwards incompatible change to change the default value of the flag, in that a system configured without the flag before the change may choose to update and then suddenly would behave differently 🤔  I do think it makes sense to recommend that folks change the default value before completely removing the flag